File: HyperFCMLive/src/main/java/io/github/howard20181/hyperos/fcmlive/theme/ThemePrefs.java
package io.github.howard20181.hyperos.fcmlive.theme;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;

import java.util.Locale;

import io.github.howard20181.hyperos.fcmlive.mcu.Scheme;

/**
 * Appearance settings: theme mode (including the AMOLED pure-black variant),
 * dynamic color with a custom seed, Material palette style, color spec version
 * and the in-app language. Stored in a private prefs file — unlike the allowlist
 * these are UI-only, so they never need to reach system_server.
 *
 * <p>Writes use {@code apply()}, not {@code commit()}: they run on the main
 * thread from click handlers, and everything that reads them back is this same
 * process, whose in-memory copy {@code apply()} updates synchronously. Only
 * {@code Prefs} needs {@code commit()}, because there the write has to be on
 * disk before the broadcast that announces it crosses into system_server.
 */
public final class ThemePrefs {

    public static final String PREFS = "fcmlive_theme";

    public static final int MODE_SYSTEM = 0;
    public static final int MODE_LIGHT = 1;
    public static final int MODE_DARK = 2;
    /** Pure-black dark variant for OLED panels. */
    public static final int MODE_AMOLED = 3;

    public static final int SPEC_2021 = 0;
    public static final int SPEC_2025 = 1;

    private static final String KEY_THEME_MODE = "theme_mode";
    private static final String KEY_PALETTE_STYLE = "palette_style";
    private static final String KEY_SPEC = "color_spec";
    private static final String KEY_DYNAMIC_COLOR = "dynamic_color";
    private static final String KEY_SEED_COLOR = "seed_color";
    /** Legacy flag from when AMOLED was a separate switch; migrated on read. */
    private static final String KEY_AMOLED = "amoled";
    private static final String KEY_LANGUAGE = "language";
    /** Follow the device language (the default: the app is not pinned). */
    public static final int LANGUAGE_SYSTEM = -1;
    /** Number of entries in {@code R.array.language_entries}. */
    public static final int LANGUAGE_COUNT = 2;

    /** Default style matches what Android's own Monet engine generates. */
    private static final int DEFAULT_STYLE = Scheme.Variant.TONAL_SPOT.ordinal();
    /** Default spec is the current Material 3 Expressive one. */
    private static final int DEFAULT_SPEC = SPEC_2025;

    private ThemePrefs() {
    }

    private static SharedPreferences prefs(Context context) {
        return context.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    public static int themeMode(Context context) {
        SharedPreferences p = prefs(context);
        int mode = p.getInt(KEY_THEME_MODE, MODE_SYSTEM);
        // Migrate the old standalone AMOLED switch into a real theme mode.
        if (mode == MODE_DARK && p.getBoolean(KEY_AMOLED, false)) {
            mode = MODE_AMOLED;
            p.edit().putInt(KEY_THEME_MODE, MODE_AMOLED).remove(KEY_AMOLED).apply();
        }
        return mode >= MODE_SYSTEM && mode <= MODE_AMOLED ? mode : MODE_SYSTEM;
    }

    public static void setThemeMode(Context context, int mode) {
        prefs(context).edit().putInt(KEY_THEME_MODE, mode).apply();
    }

    public static Scheme.Variant paletteStyle(Context context) {
        int ordinal = prefs(context).getInt(KEY_PALETTE_STYLE, DEFAULT_STYLE);
        Scheme.Variant[] values = Scheme.Variant.values();
        return ordinal >= 0 && ordinal < values.length ? values[ordinal] : values[DEFAULT_STYLE];
    }

    public static void setPaletteStyle(Context context, Scheme.Variant variant) {
        prefs(context).edit().putInt(KEY_PALETTE_STYLE, variant.ordinal()).apply();
    }

    public static int specVersion(Context context) {
        int spec = prefs(context).getInt(KEY_SPEC, DEFAULT_SPEC);
        return spec == SPEC_2021 || spec == SPEC_2025 ? spec : DEFAULT_SPEC;
    }

    public static void setSpecVersion(Context context, int spec) {
        prefs(context).edit().putInt(KEY_SPEC, spec).apply();
    }

    /** Whether colors follow the wallpaper; when off, {@link #seedColor} wins. */
    public static boolean dynamicColor(Context context) {
        return prefs(context).getBoolean(KEY_DYNAMIC_COLOR, true);
    }

    public static void setDynamicColor(Context context, boolean enabled) {
        prefs(context).edit().putBoolean(KEY_DYNAMIC_COLOR, enabled).apply();
    }

    /** Custom seed (ARGB) used while dynamic color is off; 0 means unset. */
    public static int seedColor(Context context) {
        return prefs(context).getInt(KEY_SEED_COLOR, 0);
    }

    public static void setSeedColor(Context context, int color) {
        prefs(context).edit().putInt(KEY_SEED_COLOR, color).apply();
    }

    /** Whether the resolved mode is the AMOLED pure-black variant. */
    public static boolean isAmoled(Context context) {
        return themeMode(context) == MODE_AMOLED;
    }

    /**
     * In-app language, or {@link #LANGUAGE_SYSTEM} to follow the device.
     *
     * <p>Not strictly a theme setting, but it is applied by the very same
     * configuration rewrite in {@link ThemeSupport#attach} that forces
     * light/dark, so the two are changed by one and the same recreate. Going
     * through our own preference instead of {@code LocaleManager} keeps the
     * behaviour identical on every ROM (HyperOS included) and works below the
     * per-app-language APIs.
     */
    public static int language(Context context) {
        int index = prefs(context).getInt(KEY_LANGUAGE, LANGUAGE_SYSTEM);
        return index >= 0 && index < LANGUAGE_COUNT ? index : LANGUAGE_SYSTEM;
    }

    public static void setLanguage(Context context, int index) {
        prefs(context).edit().putInt(KEY_LANGUAGE, index).apply();
    }

    /** Locale to force, or {@code null} to keep the device's own choice. */
    public static Locale locale(Context context) {
        switch (language(context)) {
            case 0:
                return Locale.SIMPLIFIED_CHINESE;
            case 1:
                return Locale.ENGLISH;
            default:
                return null;
        }
    }

    /** Resolved dark/light for the current mode (MODE_SYSTEM reads the device). */
    public static boolean isDark(Context context) {
        switch (themeMode(context)) {
            case MODE_LIGHT:
                return false;
            case MODE_DARK:
            case MODE_AMOLED:
                return true;
            default:
                int mask = context.getResources().getConfiguration().uiMode
                        & Configuration.UI_MODE_NIGHT_MASK;
                return mask == Configuration.UI_MODE_NIGHT_YES;
        }
    }
}

File: HyperFCMLive/src/main/java/io/github/howard20181/hyperos/fcmlive/theme/ThemePrefs.kt
package io.github.howard20181.hyperos.fcmlive.theme

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import java.util.Locale
import io.github.howard20181.hyperos.fcmlive.mcu.Scheme

/**
 * Appearance settings: theme mode (including the AMOLED pure-black variant),
 * dynamic color with a custom seed, Material palette style, color spec version
 * and the in-app language. Stored in a private prefs file — unlike the allowlist
 * these are UI-only, so they never need to reach system_server.
 *
 * Writes use `apply()`, not `commit()`: they run on the main
 * thread from click handlers, and everything that reads them back is this same
 * process, whose in-memory copy `apply()` updates synchronously. Only
 * `Prefs` needs `commit()`, because there the write has to be on
 * disk before the broadcast that announces it crosses into system_server.
 */
object ThemePrefs {

    const val PREFS = "fcmlive_theme"

    const val MODE_SYSTEM = 0
    const val MODE_LIGHT = 1
    const val MODE_DARK = 2
    /** Pure-black dark variant for OLED panels. */
    const val MODE_AMOLED = 3

    const val SPEC_2021 = 0
    const val SPEC_2025 = 1

    private const val KEY_THEME_MODE = "theme_mode"
    private const val KEY_PALETTE_STYLE = "palette_style"
    private const val KEY_SPEC = "color_spec"
    private const val KEY_DYNAMIC_COLOR = "dynamic_color"
    private const val KEY_SEED_COLOR = "seed_color"
    /** Legacy flag from when AMOLED was a separate switch; migrated on read. */
    private const val KEY_AMOLED = "amoled"
    private const val KEY_LANGUAGE = "language"
    /** Follow the device language (the default: the app is not pinned). */
    const val LANGUAGE_SYSTEM = -1
    /** Number of entries in `R.array.language_entries`. */
    const val LANGUAGE_COUNT = 2

    /** Default style matches what Android's own Monet engine generates. */
    private val DEFAULT_STYLE = Scheme.Variant.TONAL_SPOT.ordinal
    /** Default spec is the current Material 3 Expressive one. */
    private const val DEFAULT_SPEC = SPEC_2025

    private fun prefs(context: Context): SharedPreferences {
        return context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    }

    @JvmStatic
    fun themeMode(context: Context): Int {
        val p = prefs(context)
        var mode = p.getInt(KEY_THEME_MODE, MODE_SYSTEM)
        // Migrate the old standalone AMOLED switch into a real theme mode.
        if (mode == MODE_DARK && p.getBoolean(KEY_AMOLED, false)) {
            mode = MODE_AMOLED
            p.edit().putInt(KEY_THEME_MODE, MODE_AMOLED).remove(KEY_AMOLED).apply()
        }
        return if (mode in MODE_SYSTEM..MODE_AMOLED) mode else MODE_SYSTEM
    }

    @JvmStatic
    fun setThemeMode(context: Context, mode: Int) {
        prefs(context).edit().putInt(KEY_THEME_MODE, mode).apply()
    }

    @JvmStatic
    fun paletteStyle(context: Context): Scheme.Variant {
        val ordinal = prefs(context).getInt(KEY_PALETTE_STYLE, DEFAULT_STYLE)
        val values = Scheme.Variant.values()
        return if (ordinal in values.indices) values[ordinal] else values[DEFAULT_STYLE]
    }

    @JvmStatic
    fun setPaletteStyle(context: Context, variant: Scheme.Variant) {
        prefs(context).edit().putInt(KEY_PALETTE_STYLE, variant.ordinal).apply()
    }

    @JvmStatic
    fun specVersion(context: Context): Int {
        val spec = prefs(context).getInt(KEY_SPEC, DEFAULT_SPEC)
        return if (spec == SPEC_2021 || spec == SPEC_2025) spec else DEFAULT_SPEC
    }

    @JvmStatic
    fun setSpecVersion(context: Context, spec: Int) {
        prefs(context).edit().putInt(KEY_SPEC, spec).apply()
    }

    /** Whether colors follow the wallpaper; when off, [seedColor] wins. */
    @JvmStatic
    fun dynamicColor(context: Context): Boolean {
        return prefs(context).getBoolean(KEY_DYNAMIC_COLOR, true)
    }

    @JvmStatic
    fun setDynamicColor(context: Context, enabled: Boolean) {
        prefs(context).edit().putBoolean(KEY_DYNAMIC_COLOR, enabled).apply()
    }

    /** Custom seed (ARGB) used while dynamic color is off; 0 means unset. */
    @JvmStatic
    fun seedColor(context: Context): Int {
        return prefs(context).getInt(KEY_SEED_COLOR, 0)
    }

    @JvmStatic
    fun setSeedColor(context: Context, color: Int) {
        prefs(context).edit().putInt(KEY_SEED_COLOR, color).apply()
    }

    /** Whether the resolved mode is the AMOLED pure-black variant. */
    @JvmStatic
    fun isAmoled(context: Context): Boolean = themeMode(context) == MODE_AMOLED

    /**
     * In-app language, or [LANGUAGE_SYSTEM] to follow the device.
     *
     * Not strictly a theme setting, but it is applied by the very same
     * configuration rewrite in [ThemeSupport.attach] that forces
     * light/dark, so the two are changed by one and the same recreate. Going
     * through our own preference instead of `LocaleManager` keeps the
     * behaviour identical on every ROM (HyperOS included) and works below the
     * per-app-language APIs.
     */
    @JvmStatic
    fun language(context: Context): Int {
        val index = prefs(context).getInt(KEY_LANGUAGE, LANGUAGE_SYSTEM)
        return if (index in 0 until LANGUAGE_COUNT) index else LANGUAGE_SYSTEM
    }

    @JvmStatic
    fun setLanguage(context: Context, index: Int) {
        prefs(context).edit().putInt(KEY_LANGUAGE, index).apply()
    }

    /** Locale to force, or `null` to keep the device's own choice. */
    @JvmStatic
    fun locale(context: Context): Locale? {
        return when (language(context)) {
            0 -> Locale.SIMPLIFIED_CHINESE
            1 -> Locale.ENGLISH
            else -> null
        }
    }

    /** Resolved dark/light for the current mode (MODE_SYSTEM reads the device). */
    @JvmStatic
    fun isDark(context: Context): Boolean {
        return when (themeMode(context)) {
            MODE_LIGHT -> false
            MODE_DARK, MODE_AMOLED -> true
            else -> {
                val mask = context.resources.configuration.uiMode and
                    Configuration.UI_MODE_NIGHT_MASK
                mask == Configuration.UI_MODE_NIGHT_YES
            }
        }
    }
}
